src/main/java/picard/sam/HitsForInsert.java

```diff
@@ -194,45 +195,100 @@ public void setPrimaryAlignment(final int primaryAlignmentIndex) {
      * reads may have been filtered out.
      */
     public void coordinateByHitIndex() {
+        // null HI tag sorts after any non-null.
+        final Comparator<SAMRecord> comparator = Comparator.comparing(
+                record -> record.getIntegerAttribute(SAMTag.HI.name()),
+                Comparator.nullsLast(Comparator.naturalOrder())
+        );
         // Sort by HI value, with reads with no HI going at the end.
-        Collections.sort(firstOfPairOrFragment, comparator);
-        Collections.sort(secondOfPair, comparator);
+        firstOfPairOrFragment.sort(comparator);
+        secondOfPair.sort(comparator);
 
         // Insert nulls as necessary in the two lists so that correlated alignments have the same index
         // and uncorrelated alignments have null in the other list at the corresponding index.
         for (int i = 0; i < Math.min(firstOfPairOrFragment.size(), secondOfPair.size()); ++i) {
-            final Integer leftHi = firstOfPairOrFragment.get(i).getIntegerAttribute(SAMTag.HI.name());
-            final Integer rightHi = secondOfPair.get(i).getIntegerAttribute(SAMTag.HI.name());
-            if (leftHi != null) {
-                if (rightHi != null) {
-                    if (leftHi < rightHi) secondOfPair.add(i, null);
-                    else if (rightHi < leftHi) firstOfPairOrFragment.add(i, null);
-                    // else the are correlated
-                }
-            } else if (rightHi != null) {
-                firstOfPairOrFragment.add(i, null);
-            } else {
-                // Both alignments do not have HI, so push down the ones on the right.
-                // Right is arbitrarily picked to push down.
-                secondOfPair.add(i, null);
+            final Integer hiTagValueOfFirst = firstOfPairOrFragment.get(i).getIntegerAttribute(SAMTag.HI.name());
+            final Integer hiTagValueOfSecond = secondOfPair.get(i).getIntegerAttribute(SAMTag.HI.name());
+
+            if (nonNull(hiTagValueOfFirst) && nonNull(hiTagValueOfSecond)) {
```

Comment:
What is the rationale for using these functions instead of comparing against `null` directly? My understanding is that these methods are provided so we have convenient lambda forms for streaming operations, etc. Using them inline like this doesn't look like an improvement to me.

Reply:
I prefer to use functional style instead of procedural even if it could decrease performance a little bit because it usually increases code readability. Especially for functions of Java standard library. I could revert all cases when I use static import if it is the concern. So is it critical?

Reply:
Not sure what you mean about functional vs procedural... one form is an expression consisting of a function call, the other is a binary expression using the`!=` operator. Both are equally functional and procedural, as I understand these terms.

Readability is a subjective argument; to me, using `!= null` is more readable. Given that it requires a static import to avoid making the code much wordier than the original, I don't think it's worth the change.

Reply:
Done [here](https://github.com/broadinstitute/picard/pull/932/commits/96bc12a053504430c0ae1c1aafcea8667924648a).

Reply:
Speaking about operators vs functions, I have always thought that since Java doesn't allow to make partial application of an operator it is convenient to use functions in such cases. Let's imagine we have `equals` function that takes two objects and checks their equality. So `isNull` function is just a partial application of `equals` function with fixed `null` as one of the arguments. Then `nonNull` is negated `isNull`. So that way, for the sake of consistency it is easier to use functions instead of operators in Java not only in streams but everywhere in mine opinion. Is it a little bit excessively? :smile:

Reply:
It sounds like you have a more enlightened view of functional programming in Java than I do. :)

Thanks for the explanation, I now understand the distinction you're making between expressions and functions, and why it's more consistent to use functions everywhere. I can't say I'm convinced but I'll definitely be giving it more thought.